File: app/src/main/java/ua/fvadevand/reminderstatusbar/adapters/ReminderAdapter.java
package ua.fvadevand.reminderstatusbar.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.List;
import java.util.Objects;

import ua.fvadevand.reminderstatusbar.R;
import ua.fvadevand.reminderstatusbar.data.models.Reminder;
import ua.fvadevand.reminderstatusbar.utilities.IconUtils;

public class ReminderAdapter extends RecyclerView.Adapter<ReminderAdapter.ReminderViewHolder> {

    private List<Reminder> mReminderList;
    private OnReminderClickListener mListener;

    public ReminderAdapter(OnReminderClickListener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public ReminderViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.reminder_list_item, parent, false);
        return new ReminderViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ReminderViewHolder holder, int position) {
        holder.bind(mReminderList.get(position));
    }

    @Override
    public int getItemCount() {
        return mReminderList == null ? 0 : mReminderList.size();
    }

    public void setReminderList(List<Reminder> newList) {
        if (newList == null) return;
        if (mReminderList == null) {
            mReminderList = newList;
            notifyDataSetChanged();
        } else {
            ReminderDiffUtilCallback diffUtilCallback = new ReminderDiffUtilCallback(mReminderList, newList);
            DiffUtil.DiffResult diffUtilResult = DiffUtil.calculateDiff(diffUtilCallback);
            mReminderList.clear();
            mReminderList.addAll(newList);
            diffUtilResult.dispatchUpdatesTo(this);
        }
    }

    public interface OnReminderClickListener {
        void onReminderClick(long id);
    }

    class ReminderDiffUtilCallback extends DiffUtil.Callback {

        private List<Reminder> mOldList;
        private List<Reminder> mNewList;

        ReminderDiffUtilCallback(List<Reminder> oldList, List<Reminder> newList) {
            mOldList = oldList;
            mNewList = newList;
        }

        @Override
        public int getOldListSize() {
            return mOldList.size();
        }

        @Override
        public int getNewListSize() {
            return mNewList.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return mOldList.get(oldItemPosition).getId() == mNewList.get(newItemPosition).getId();
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return Objects.equals(mOldList.get(oldItemPosition), mNewList.get(newItemPosition));
        }
    }

    class ReminderViewHolder extends RecyclerView.ViewHolder {

        ImageView mIconView;
        TextView mTitleView;
        TextView mTextView;
        TextView mDateView;

        ReminderViewHolder(View itemView) {
            super(itemView);
            mIconView = itemView.findViewById(R.id.iv_item_icon);
            mTitleView = itemView.findViewById(R.id.tv_item_title);
            mTextView = itemView.findViewById(R.id.tv_item_text);
            mDateView = itemView.findViewById(R.id.tv_item_date);
            itemView.setOnClickListener(v -> {
                int position = getAdapterPosition();
                if (mListener != null && position != RecyclerView.NO_POSITION) {
                    mListener.onReminderClick(mReminderList.get(position).getId());
                }
            });
        }

        void bind(Reminder reminder) {
            Context context = itemView.getContext();
            mIconView.setImageResource(IconUtils.getIconResId(context, reminder.getIconName()));
            mTitleView.setText(reminder.getTitle());
            mTextView.setText(reminder.getText());
        }
    }
}

File: app/src/main/java/ua/fvadevand/reminderstatusbar/adapters/ReminderAdapter.kt
package ua.fvadevand.reminderstatusbar.adapters

import android.animation.AnimatorSet
import android.animation.ArgbEvaluator
import android.animation.ValueAnimator
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.core.animation.doOnEnd
import androidx.core.animation.doOnStart
import androidx.core.view.isVisible
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.RecyclerView
import ua.fvadevand.reminderstatusbar.R
import ua.fvadevand.reminderstatusbar.adapters.utils.ReminderDiffUtil
import ua.fvadevand.reminderstatusbar.data.models.Reminder
import ua.fvadevand.reminderstatusbar.data.models.ReminderStatus
import ua.fvadevand.reminderstatusbar.databinding.ItemReminderBinding
import ua.fvadevand.reminderstatusbar.decorators.SwipeToEditOrDeleteCallback
import ua.fvadevand.reminderstatusbar.listeners.OnReminderInteractListener
import ua.fvadevand.reminderstatusbar.utils.getNotificationTime
import ua.fvadevand.reminderstatusbar.utils.setImageResourceName

class ReminderAdapter(
    private val highlightColor: Int,
    private val listener: OnReminderInteractListener?
) : RecyclerView.Adapter<ReminderAdapter.ReminderViewHolder>(),
    SwipeToEditOrDeleteCallback.SwipeableAdapter {

    private val differ = AsyncListDiffer(this, ReminderDiffUtil())
    private val reminders get() = differ.currentList
    private var highlightPosition = -1
    private val rgbEvaluator by lazy { ArgbEvaluator() }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ReminderViewHolder {
        val itemBinding = ItemReminderBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return ReminderViewHolder(itemBinding)
    }

    override fun onBindViewHolder(holder: ReminderViewHolder, position: Int) {
        holder.bind(reminders[position])
        if (position == highlightPosition) {
            highlightPosition = -1
            holder.highlight()
        }
    }

    override fun getItemViewType(position: Int): Int {
        return R.layout.item_reminder
    }

    override fun getItemCount(): Int {
        return reminders.size
    }

    override fun allowSwipe(position: Int): Boolean {
        return true
    }

    override fun editItem(position: Int) {
        if (isValidPosition(position)) {
            notifyItemChanged(position)
            listener?.onReminderEdit(getReminderByPosition(position).id)
        }
    }

    override fun deleteItem(position: Int) {
        if (isValidPosition(position)) {
            listener?.onReminderDelete(getReminderByPosition(position))
        }
    }

    fun submitReminders(newList: List<Reminder>, onComplete: () -> Unit) {
        differ.submitList(newList, onComplete)
    }

    fun getReminderIndex(reminderId: Long): Int {
        return reminders.indexOfFirst { it.id == reminderId }
    }

    fun highlightItem(position: Int) {
        highlightPosition = position
        notifyItemChanged(position)
    }

    private fun isValidPosition(position: Int) = position in 0 until itemCount

    private fun getReminderByPosition(position: Int) = reminders[position]

    inner class ReminderViewHolder(
        private val itemBinding: ItemReminderBinding
    ) : RecyclerView.ViewHolder(itemBinding.root) {

        private var animator: AnimatorSet? = null

        init {
            itemBinding.root.setOnClickListener {
                val position = bindingAdapterPosition
                if (isValidPosition(position)) {
                    listener?.onReminderClick(getReminderByPosition(position).id)
                }
            }
        }

        fun bind(reminder: Reminder) {
            itemBinding.ivItemReminderIcon.setImageResourceName(reminder.iconName)
            itemBinding.tvItemReminderTitle.text = reminder.title
            val reminderText = reminder.text
            if (reminderText.isNullOrEmpty()) {
                itemBinding.tvItemReminderText.isVisible = false
            } else {
                itemBinding.tvItemReminderText.isVisible = true
                itemBinding.tvItemReminderText.text = reminderText
            }
            if (reminder.timestamp > System.currentTimeMillis() &&
                reminder.status != ReminderStatus.PAUSED
            ) {
                itemBinding.tvItemReminderDate.isVisible = true
                itemBinding.tvItemReminderDate.text =
                    itemView.context.getNotificationTime(reminder.timestamp)
            } else {
                itemBinding.tvItemReminderDate.isVisible = false
            }
            itemBinding.ivItemReminderStatus.setImageResource(
                ReminderStatus.getIconResIdByStatus(reminder.status)
            )
        }

        fun highlight() {
            animator?.end()
            val backgroundDrawable = itemView.background as? GradientDrawable ?: return
            val backgroundColor = backgroundDrawable.color?.defaultColor ?: return
            backgroundDrawable.mutate()
            val inAnimation = ValueAnimator.ofFloat(0F, 1F).apply {
                duration = 750
                addUpdateListener {
                    backgroundDrawable.setColor(
                        rgbEvaluator.evaluate(
                            it.animatedFraction,
                            backgroundColor,
                            highlightColor
                        ) as Int
                    )
                }
            }

            val outAnimation = ValueAnimator.ofFloat(0F, 1F).apply {
                duration = 750
                addUpdateListener {
                    backgroundDrawable.setColor(
                        rgbEvaluator.evaluate(
                            it.animatedFraction,
                            highlightColor,
                            backgroundColor
                        ) as Int
                    )
                }
            }

            animator = AnimatorSet().apply {
                playSequentially(inAnimation, outAnimation)
                doOnStart { itemView.setHasTransientState(true) }
                doOnEnd {
                    animator = null
                    itemView.setHasTransientState(false)
                }
                start()
            }
        }

    }

}
